qemu/qemu.go

```diff
@@ -646,6 +647,400 @@ const (
+	// Available values are: on, off
+	CopyOnRead DriveState
+
+	// DetectZeroes enables the automatic conversion of plain zero writes by the OS to driver specific optimized zero write
+	// Available values are: off, on, unmap
+	DetectZeroes DriveState
+}
+
+// DriveInterface specifies the interface where the drive is connected
+type DriveInterface string
+
+const (
+	// DriveIde ide interface
+	DriveIde DriveInterface = "ide"
+
+	// DriveScsi scsie interface
+	DriveScsi DriveInterface = "scsi"
+
+	// DriveSd sd interface
+	DriveSd DriveInterface = "sd"
+
+	// DriveMtd mtd interface
+	DriveMtd DriveInterface = "mtd"
+
+	// DriveFloppy floppy interface
+	DriveFloppy DriveInterface = "floppy"
+
+	// DrivePflash pflash interface
+	DrivePflash DriveInterface = "pflash"
+
+	// DriveVirtio virtio interface
+	DriveVirtio DriveInterface = "virtio"
+)
+
+// DriveMedia specifies the drive media
+type DriveMedia string
+
+const (
+	// DriveCdrom cdrom media
+	DriveCdrom DriveMedia = "cdrom"
+
+	// DriveDisk disk media
+	DriveDisk DriveMedia = "disk"
+)
+
+// DriveCacheMode specifies the cache mode
+type DriveCacheMode string
+
+const (
+	// DriveCacheNone none cache
+	DriveCacheNone DriveCacheMode = "none"
+
+	// DriveCacheWriteback writeback cache
+	DriveCacheWriteback DriveCacheMode = "writeback"
+
+	// DriveCacheUnsafe unsafe cache
+	DriveCacheUnsafe DriveCacheMode = "unsafe"
+
+	// DriveCacheDirectsync directsync cache
+	DriveCacheDirectsync DriveCacheMode = "directsync"
+
+	// DriveCacheWritethrough writethrough cache
+	DriveCacheWritethrough DriveCacheMode = "writethrough"
+)
+
+// DriveAIO specifies the Kernel Asynchronous I/O
+type DriveAIO string
+
+const (
+	// DriveAIOThreads pthread based disk I/O
+	DriveAIOThreads DriveAIO = "threads"
+
+	// DriveAIONative native linux AIO
+	DriveAIONative DriveAIO = "native"
+)
+
+// DriveState specifies the state for a drive feature
+type DriveState string
+
+const (
+	// DriveStateOff disable a drive feature
+	DriveStateOff DriveState = "off"
+
+	// DriveStateOn enable a drive feature
+	DriveStateOn DriveState = "on"
+
+	// DriveStateUnmap same as on
+	DriveStateUnmap DriveState = "unmap"
+
+	// DriveStateIgnore same as off
+	DriveStateIgnore DriveState = "ignore"
+)
+
+func validDriveString(value interface{}, expectedValues ...interface{}) bool {
+	if reflect.DeepEqual(value, "") {
+		return true
+	}
+
+	for _, e := range expectedValues {
+		if reflect.DeepEqual(value, e) {
+			return true
+		}
+	}
+
+	return false
+}
+
+func (drive Drive) validDriveStrings() bool {
+	if drive.File == "" {
+		return false
+	}
+
+	if drive.Index != "" {
+		if _, err := strconv.ParseUint(drive.Index, 10, 32); err != nil {
+			return false
+		}
+	}
+
+	if drive.Addr != "" {
+		if drive.If != DriveVirtio {
+			return false
+		}
+	}
+
+	return validDriveString(drive.If, DriveIde, DriveScsi, DriveSd, DriveMtd, DriveFloppy, DrivePflash, DriveVirtio) &&
```

Comment:
I'm wondering if reflection is overkill here.  Couldn't we just have a IsValid method for all the types except DriveState and then one specific method for DriveState that allows us to specify different sets of assignable values?  This way we wouldn't need to use reflection.  Alternatively, we could define three different types for DriveState to represent the permissable sets of options.

Reply:
fixed